### PR TITLE
check-by-name: Refactor to prepare for enforcing `pkgs/by-name`, make `--base` required

### PR DESCRIPTION
**Pull request by infinisil**
## Description of changes

This is a preparation PR for https://github.com/NixOS/nixpkgs/pull/275539, including only internal improvements to the code and comments, without changing any functionality. This makes the diff for the above PR much smaller.

The only somewhat interesting change is that `--base` is now required for `nixpkgs-check-by-name`, which was already prepared for in https://github.com/NixOS/nixpkgs/pull/274591 by always passing the flag in CI.

This PR is best reviewed commit-by-commit.

## Things done

- [x] Built and tested on `x86_64-linux`

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/test/nixpkgs-check-by-name/tests/empty-base/pkgs/by-name/README.md
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/tests/empty-base/pkgs/by-name/README.md:1]: I'm assuming this is a sentinel file.

Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/tests/empty-base/pkgs/by-name/README.md:1]: Yeah, just so that git can track that the directory exists. Marginally easier than changing the code to handle a missing directory

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs [hunk 1]
```diff
@@ -6,33 +6,48 @@ use std::path::Path;
...
 
 /// Attribute set of this structure is returned by eval.nix
 #[derive(Deserialize)]
-struct AttributeInfo {
-    variant: AttributeVariant,
+enum ByNameAttribute {
+    /// The attribute doesn't exist at all
+    Missing,
+    Existing(AttributeInfo),
+}
```
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/eval.rs:19]: These types are much clearer than what came before.

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs [hunk 2]
```diff
@@ -103,74 +118,88 @@ pub fn check_values(
...
+            use AttributeInfo::*;
+            use ByNameAttribute::*;
+            use CallPackageVariant::*;
+
+            let check_result = match attribute_value {
+                Missing => NixpkgsProblem::UndefinedAttr {
+                    relative_package_file: relative_package_file.clone(),
+                    package_name: attribute_name.clone(),
+                }
+                .into(),
```
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/eval.rs:141]: This looks like it's implementing `IntoNixpkgsProblem` on `ByNameAttribute`.

The same thing holds for the cases below on their specific types.

File: pkgs/test/nixpkgs-check-by-name/src/eval.nix
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/eval.nix:1]: I only lightly reviewed this file. I didn't see anything that caused me to believe that it's causing the tests to be misleading, and appears to be monkey-patching or mocking callPackage well enough.

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs [hunk 3]
```diff
@@ -103,74 +118,88 @@ pub fn check_values(
         anyhow::bail!("Failed to run command {command:?}");
     }
     // Parse the resulting JSON value
-    let actual_files: HashMap<String, AttributeInfo> = serde_json::from_slice(&result.stdout)
+    let attributes: Vec<(String, ByNameAttribute)> = serde_json::from_slice(&result.stdout)
         .context(format!(
             "Failed to deserialise {}",
             String::from_utf8_lossy(&result.stdout)
         ))?;
```
Comment (philiptaron) [inline pkgs/test/nixpkgs-check-by-name/src/eval.rs:125]: Here and throughout the crate: use `anyhow::Context::with_context` instead of `context` to avoid allocating these error strings in the non-error case.

**Review by philiptaron — APPROVED** [review]
OK, I took a look at the Rust code here, and skimmed the test-only Nix code.

I have three non-blocking refactoring suggestions, in descending order of "want to see":

1. Simplify the traits in `ratchet.rs` and use them more broadly.
2. Use `anyhow::Context::with_context` to avoid allocating error strings in the non-error case.
3. Minor typo and comment fixes as noted.

There's a slightly larger refactoring that I see that's not directly related to the changes in this PR. In main, everywhere the `error_writer` is used could instead use a `ratchet::Nixpkgs` that's cased on its trivial success.

File: pkgs/test/nixpkgs-check-by-name/src/eval.nix [hunk 4]
```diff
@@ -1,82 +1,86 @@
...
+  overlay = final: prev: {
+
+    # Information for attributes defined using `callPackage`
     callPackage = fn: args:
-      let
-        result = super.callPackage fn args;
-        variantInfo._attributeVariant = {
-          # These names are used by the deserializer on the Rust side
-          CallPackage.path =
+      addVariantInfo (prev.callPackage fn args) {
```
Comment (Ericson2314) [inline pkgs/test/nixpkgs-check-by-name/src/eval.nix:22]: uhhh how is this in scope?

**Review by Ericson2314 — APPROVED** [review]
LGTM!

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs (hunk 2 above)
Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/src/eval.rs:141]: Not entirely because in some cases there won't be an error.

File: pkgs/test/nixpkgs-check-by-name/src/ratchet.rs [hunk 5]
```diff
@@ -57,29 +46,61 @@ impl Package {
...
+    /// The ratchet is tight, it can't be tightened any further.
+    /// This is either because we already use the latest state, or because the ratchet isn't
+    /// relevant.
+    Tight,
+}
+
+/// A trait for a specific ratchet check for an attribute.
+trait AttributeRatchet: Sized {
+    /// What error to produce in case the ratchet went in the wrong direction
+    fn to_error(name: &str, context: &Self, existed_before: bool) -> NixpkgsProblem;
```
[the remark this answers is not in the document]
Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/src/ratchet.rs:66]: It looks like making it consume `Self` is a bit trickier, since the values come from a HashMap originally. And not sure what you mean with `context` being `self`!

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs (hunk 3 above)
Comment (infinisil) [reply pkgs/test/nixpkgs-check-by-name/src/eval.rs]: Done throughout the codebase in https://github.com/NixOS/nixpkgs/pull/278805/commits/a1db0cdf9bb5e90727eb1c455da448fb21bab91e :)

**Comment by infinisil** [comment]
Feedback addressed, I'll merge once CI passes since the feedback isn't blocking :)

File: pkgs/test/nixpkgs-check-by-name/src/eval.rs (hunk 2 above)
Comment (philiptaron) [reply pkgs/test/nixpkgs-check-by-name/src/eval.rs:145]: > Not entirely because in some cases there won't be an error.

Agreed. I took a swing at this (PR not sent out) and ran into the same thing. There's some fine tuning that could be done for sure but I think the trait-based approach I suggested isn't right.

**Review by philiptaron — APPROVED** [review]
Ship it.
